parsers/jscript.c

```diff
@@ -46,6 +47,52 @@
 #define deleteToken(t) (objPoolPut (TokenPool, (t)))
 
 /*
+ * Debugging
+ *
+ * Uncomment this to enable extensive debugging to stderr in jscript code.
+ * Please note that TRACING_ENABLED should be #defined in main/trace.h
+ * for this to work.
+ *
+ */
+//#define JSCRIPT_DEBUGGING_ENABLED 1
+
+#if defined(DO_TRACING) && defined(JSCRIPT_DEBUGGING_ENABLED)
+	#define JSCRIPT_DO_DEBUGGING
+#endif
+
+#ifdef JSCRIPT_DO_DEBUGGING
+
+#define JSCRIPT_DEBUG_ENTER() TRACE_ENTER()
+#define JSCRIPT_DEBUG_LEAVE() TRACE_LEAVE()
+
+#define JSCRIPT_DEBUG_ENTER_TEXT(_szFormat,...) \
+	TRACE_ENTER_TEXT(_szFormat,## __VA_ARGS__)
+
+#define JSCRIPT_DEBUG_LEAVE_TEXT(_szFormat,...) \
+	TRACE_LEAVE_TEXT(_szFormat,## __VA_ARGS__)
+
+#define JSCRIPT_DEBUG_PRINT(_szFormat,...) \
+	TRACE_PRINT(_szFormat,## __VA_ARGS__)
+
+#define JSCRIPT_DEBUG_ASSERT(_condition,_szFormat,...) \
+	TRACE_ASSERT(_condition,_szFormat,## __VA_ARGS__)
+
+#else //!JSCRIPT_DO_DEBUGGING
+
+#define JSCRIPT_DEBUG_ENTER() do { } while(0)
+#define JSCRIPT_DEBUG_LEAVE() do { } while(0)
+
+#define JSCRIPT_DEBUG_ENTER_TEXT(_szFormat,...) do { } while(0)
+#define JSCRIPT_DEBUG_LEAVE_TEXT(_szFormat,...) do { } while(0)
+
+#define JSCRIPT_DEBUG_PRINT(_szFormat,...) do { } while(0)
+
+#define JSCRIPT_DEBUG_ASSERT(_condition,_szFormat,...) do { } while(0)
+
+#endif //!JSCRIPT_DO_DEBUGGING
+
+
+/*
```

Comment:
What about instead having a symbol a module can define before including *trace.h* that select whether the macros are enabled?  Could even add a prefix or something if we wanted.

I don't know, like
```C
#define DO_MODULE_TRACING JS_DO_DEBUGGING
#include "trace.h"
```
and in *trace.h* do something like
```C
#if defined(DO_TRACING) && (! defined (DO_MODULE_TRACING) || DO_MODULE_TRACING)
…
```
so if a file defined `DO_MODULE_TRACING` before, it conditionally enables `TRACE_…`, but leaves things not defining that as it is currently.

Though, that sounds like maybe simply
```
#if ! JS_DO_DEBUGGING
# undef DO_TRACING
#endinf
#include "trace.h"
```
would be the same.

Anyway, I somewhat agree that granularity would be nice, but also that @masatake is right and this is too verbose and should be made more generic.  As one can't declare a macro inside a macro, direct use of the `TRACE_*` calls in the modules would be needed I guess, so having the granularity in *trace.h*.

---
and agreed on the separate commit, would be nicer